### PR TITLE
Allow the register allocator to use all x64 registers

### PR DESCRIPTION
**Pull request by jspam**

### CODE REVIEW

File: src/main/scala/mjis/PhiCodeGenerator.scala [hunk 1]
```diff
@@ -88,29 +100,36 @@ class PhiCodeGenerator(inputPhi: Map[Operand, Operand], cycleTempRegNr: Int, mem
...
   }
 
-  def getInstructions(): Seq[Instruction] = {
-    buildCyclesAndRoots()
+  def neededTempRegs = {
+    def isMemOp(op: Operand) = op.isInstanceOf[ActivationRecordOperand] || op.isInstanceOf[AddressOperand]
+    Seq(
+      inputPhi.exists(t => isMemOp(t._1) && isMemOp(t._2)),
+      cycles.nonEmpty
+    ).count(_ == true)
```
Comment (Kha) [inline src/main/scala/mjis/PhiCodeGenerator.scala]: *ahem* :)

File: src/main/scala/mjis/PhiCodeGenerator.scala [hunk 2]
```diff
@@ -88,29 +104,39 @@ class PhiCodeGenerator(inputPhi: Map[Operand, Operand], cycleTempRegNr: Int, mem
...
   }
 
-  def getInstructions(): Seq[Instruction] = {
-    buildCyclesAndRoots()
+  lazy val neededTempRegs = {
+    def isMemOp(op: Operand) = op.isInstanceOf[ActivationRecordOperand] || op.isInstanceOf[AddressOperand]
+    if (cycles.nonEmpty)
+      // Need one temp register for the cycle,
+      // and another one if there are memory-to-memory moves in the cycle.
+      if (cycles.exists(_.sliding(2).exists(s => isMemOp(s(0)) && isMemOp(s(1))))) 2 else 1
```
Comment (Kha) [inline src/main/scala/mjis/PhiCodeGenerator.scala]: What if the first and last operand are mem operands?

Comment (Kha) [inline src/main/scala/mjis/PhiCodeGenerator.scala]: Also, IndexOutOfBounds if only the last one is a mem operand

Comment (jspam) [inline src/main/scala/mjis/PhiCodeGenerator.scala]: They are moved to/from the temp register respectively, so no mem-mem-move is generated there.

**Comment by TehMillhouse** [comment]
Wow, impressive piece of work. Quite a chunk, but a good one. Let's see what speedcenter says to this. :smile:

**Pull request merged**